fips: improve PCT for DH

xnox opened the pull request:
In generate_key() ensure public key is checked, and pairwise check is
performed.

Based on https://gitlab.com/redhat/centos-stream/rpms/openssl/-/blob/c9s/0044-FIPS-140-3-keychecks.patch by @beldmit

Forward port to current codebase, use return errors rather than aborting the library.

IG 10.3.A additional comment 1 on page 87 (10/23/2024 edition of FIPS 140-3 IG) states the following:

> Though not a CAST, a pairwise consistency test (PCT) shall be conducted for every generated public and private key pair for the applicable approved algorithm (per ISO/IEC 19790:2012 Section 7.10.3.3). At minimum, the PCT that is required by the underlying algorithm standard (e.g., SP 800-56Arev3 Section 5.6.2.1.4 option ‘b’ or SP 800-56Brev2 Section 6.4.1.1 option ‘2.’) shall be performed, which will satisfy the PCT requirement in either TE10.35.01 for key transport, TE10.35.02 for signatures, or TE10.35.03 for key agreement.

The SP 800-56Arev3 Section 5.6.2.1.4(b) is to validate public parameters & recompute and compare the public key a second time. This is now implemented with calls to `DH_check_pub_key()` and ossl_dh_check_pairwise().

Co-authored-by: Dmitry Belyavskiy <dbelyavs@redhat.com>

Signed-off-by: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>

xnox committed:
fips: dh: perform PCT in generate_key()

ISO/IEC 19790 Section 7.10.3.3 requires a pair-wise consistentcy test
to be performed for every generated public and private key pair.

In generate_key() ensure public key is checked, and pairwise check is
performed.

Based on https://gitlab.com/redhat/centos-stream/rpms/openssl/-/blob/c9s/0044-FIPS-140-3-keychecks.patch by @beldmit

Forward port to current codebase, use return errors rather than
aborting the library.

IG 10.3.A additional comment 1 on page 87 (10/23/2024 edition of FIPS
140-3 IG) states the following:

"Though not a CAST, a pairwise consistency test (PCT) shall be
conducted for every generated public and private key pair for the
applicable approved algorithm (per ISO/IEC 19790:2012 Section
7.10.3.3). At minimum, the PCT that is required by the underlying
algorithm standard (e.g., SP 800-56Arev3 Section 5.6.2.1.4 option ‘b’
or SP 800-56Brev2 Section 6.4.1.1 option ‘2.’) shall be performed,
which will satisfy the PCT requirement in either TE10.35.01 for key
transport, TE10.35.02 for signatures, or TE10.35.03 for key
agreement."

The SP 800-56Arev3 Section 5.6.2.1.4(b) is to validate parameters &
recompute and compare the public key a second time. This is now
implemented with calls to `DH_check_pub_key()` and
ossl_dh_check_pairwise().

Co-authored-by: Dmitry Belyavskiy <dbelyavs@redhat.com>

Signed-off-by: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>

beldmit commented:
I think these patches were also co-authored with several other Red Hat Developers.

xnox commented:
> According to the [history of that patch](https://gitlab.com/redhat/centos-stream/rpms/openssl/-/commits/c9s/0044-FIPS-140-3-keychecks.patch), I'm the other co-author. I only made a minor compiler warning fix, and added an `abort()` call due to disagreement documented in #22506 that's probably not appropriate to include here (and also doesn't seem to be included).

abort() is a valid way to ensure a cryptographic module stops producing output. In this codebase, module enters error state at appropriate times. If needed, we can ensure provider enters errors state upon PCT - which has been added to all relevant APIs in https://github.com/openssl/openssl/pull/25580 . However, imho PCT is a recoverable error - as likely reason for failure is a temporary memory failure, rather than something requiring a module reset.

beldmit commented:
I agree that the current code base makes no sense to introduce the `abort()` call.

t8m commented:
_<fips_rant>_ Another reason why FIPS regulations are crazy. This will just double the overhead of DH and ECDH key exchange on the server. But I assume performance and FIPS is kind of oxymoron and we will have to take this. </fips_rant>_

xnox commented:
> _<fips_rant>_ Another reason why FIPS regulations are crazy. This will just double the overhead of DH and ECDH key exchange on the server. But I assume performance and FIPS is kind of oxymoron and we will have to take this. _</fips_rant>_

I tried hard pushing back on the grounds that it is really expensive operation. But the ISO standard section is just that sentence saying one must do PCT upon any key generation. And it is hard to argue that DH & keygen are not generating new keys.

And really it is ISO text requirement, which FIPS references. So please do have a more international rant. As said ISO standard likely making its way to all other jurisdictions too, beyond just FIPS.

paulidale commented on providers/implementations/keymgmt/ec_kmgmt.c:
```diff
@@ -993,9 +993,18 @@ struct ec_gen_ctx {
     OSSL_FIPS_IND_DECLARE
 };
 
+#ifdef FIPS_MODULE
```
These should live in a prov header.

xnox replied:
ack, but will do this in a separate PR.

paulidale commented on providers/implementations/keymgmt/rsa_kmgmt.c:
```diff
@@ -447,6 +448,12 @@ static int rsa_gencb(int p, int n, BN_GENCB *cb)
     return gctx->cb(params, gctx->cbarg);
 }
 
+#ifdef FIPS_MODULE
```
These should live in a prov header.

xnox replied:
Ack, but probably this is not needed.

slontis commented on providers/implementations/keymgmt/ec_kmgmt.c:
```diff
@@ -1332,6 +1345,17 @@ static void *ec_gen(void *genctx, OSSL_CALLBACK *osslcb, void *cbarg)
         ret = ret && ossl_ec_set_ecdh_cofactor_mode(ec, gctx->ecdh_mode);
+#ifdef FIPS_MODULE
+    /* Pairwise consistency test */
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0
```
Doesn't ec_generate_key() already do a PCT?

xnox replied:
> crypto/ec/ec_key.c:ecdsa_keygen_pairwise_test()
> As also required by CAVP FAQ GEN.2, this PCT shall perform hash+sign digital signature operations using an atomic function i.e., cannot rely on ECDSA_do_sign and ECDSA_do_verify which do not perform the hashing.

xnox replied:
So my patch is wrong, and instead it should be fixing ecdsa_keygen_pariwise_test to hash a message and sign it. Rather than signing a digest of zeros.

beldmit replied:
The original patch was written the way it uses the combined callbacks (so that's why some functions stopped being static). I don't remember whether the pairwise test was called at that moment.

xnox replied:
Ack, but will make changes to ec_generate_key in a separate PR.

slontis commented on providers/implementations/keymgmt/rsa_kmgmt.c:
```diff
@@ -628,6 +639,14 @@ static void *rsa_gen(void *genctx, OSSL_CALLBACK *osslcb, void *cbarg)
 
+#ifdef FIPS_MODULE
+    /* Pairwise consistency test */
+    if (do_rsa_pct(gctx->prov_rsa_ctx, "sha256", rsa_tmp) != 1) {
```
rsa_keygen() does a pct??

xnox replied:
yes rsa_keygen_pairwise_test looks good. Let me drop RSA changes.

slontis commented:
As far as I am aware we are already doing the relevant PCT tests in the low level gen calls.

xnox commented:
> As far as I am aware we are already doing the relevant PCT tests in the low level gen calls.
> 

Let me check again. It is done with Ed ones, but let me recheck RSA and Ec curves.

xnox commented:
I don't see any in Ed curves.

xnox commented:
@paulidale @slontis @beldmit 

I am reducing the scope of this PR:
- rsa changes not needed, dropped (although I do think it would be better to switch to OAEP instead of RAW test in rsa_gen.c, but low priority)
- ec changes as initially proposed dropped. But will be reintroduced in a separate PR - in essence the ED25519 pairwise test is correct and uses a full message; so will upgrade ECDSA test to do the same in the similar manner.

What now remains:
- changes to DH to ECDH to verify pairwise, peer, public keys

slontis commented:
I am not sure we should be doing this either..
It checks the public key of the peer already in EVP_PKEY_derive_set_peer_ex().
The assurance of the public key on the local side does not always need to be checked, depending on if the key is trusted or not.

xnox commented:
> I am not sure we should be doing this either.. It checks the public key of the peer already in EVP_PKEY_derive_set_peer_ex(). The assurance of the public key on the local side does not always need to be checked, depending on if the key is trusted or not.

Ack, let me trace all that.

xnox commented:
> I am not sure we should be doing this either.. It checks the public key of the peer already in EVP_PKEY_derive_set_peer_ex(). The assurance of the public key on the local side does not always need to be checked, depending on if the key is trusted or not.

Indeed all that is already checked. The only hunks that are left then are just this:

https://github.com/openssl/openssl/pull/25863/files

which of dubious benefit; as the public key which may have been passed into the function; has already been discarded and recomputed fresh.

Possibly, in case of not generating private key; and when a public key was passed in too....... we could validate them pairwise and return an error from the function; instead of silently correcting public key to the correct one. But I don't understand what purpose that could serve.

xnox commented:
> Might be good to review the RH code to see if this PR was necessary there. It may have been relevant when it was written, but it doesnt look like it is needed now. (I dont think you want to do the pairwise tests twice :)) All of the Pairwise FIPS tests really annoy me, as they are completely pointless for a software module. If they ever fail then something is drastically wrong. If algorithms are implemented using the standards for key gen then that should actually be sufficient.

Updated the PR description and commit with direct quotes from FIPS 140-3 IG. A pairwise test is needed; if at generation time usage is not known "at minimum" one test needs to be performed..... which indeed is to recalculate the public key a second time and compare it. As pointless as it is. Everyone I spoke to agreed that this is silly, yet still necessary. And this silly check to recompute public key a second time, is the smallest 'at minimum' implementation.

xnox committed:
fips: dh: also revalidate remote key in ossl_dh_compute_key

Recheck remote public key compliant to SP800-56A Rev 3 Section 5.6.2.2
to prevent possibility of remote key not being check by the FIPS
module.

xnox committed:
fips: dh: set FIPS module into error state when PCT fails

If pairwise test fails in generate_key() ensure the module enters
error state, like in all other PCT failure scenarios.

t8m commented on crypto/dh/dh_key.c:
```diff
@@ -369,8 +386,23 @@ static int generate_key(DH *dh)
     if (!ossl_dh_generate_public_key(ctx, dh, priv_key, pub_key))
         goto err;
 
+#ifdef FIPS_MODULE
+    if (DH_check_pub_key(dh, pub_key, &validate) <= 0) {
+        ERR_raise(ERR_LIB_DH, DH_R_CHECK_PUBKEY_INVALID);
+        goto err;
+    }
+#endif
+
     dh->pub_key = pub_key;
     dh->priv_key = priv_key;
+#ifdef FIPS_MODULE
+    if (ossl_dh_check_pairwise(dh) <= 0) {
+        ossl_set_error_state(OSSL_SELF_TEST_TYPE_PCT);
+        ERR_raise(ERR_LIB_PROV, DH_R_CHECK_PUBKEY_INVALID);
+        goto err;
+    }
+#endif
+
     dh->dirty_cnt++;
```
Why both pairwise and pub key check would be needed? That does not make sense at all to me.

t8m commented on crypto/dh/dh_key.c:
```diff
@@ -60,6 +67,13 @@ int ossl_dh_compute_key(unsigned char *key, const BIGNUM *pub_key, DH *dh)
         return 0;
     }
 
+#ifdef FIPS_MODULE
+    if (DH_check_pub_key(dh, pub_key, &validate) <= 0) {
+        ERR_raise(ERR_LIB_DH, DH_R_CHECK_PUBKEY_INVALID);
+        return 0;
+    }
+#endif
+
```
This is a very expensive operation which is really unnecessary for the known safe prime DH parameters and furthermore you do not check the validate value so you do not really reject bad pub keys at all.

You should instead use ossl_ffc_validate_public_key_partial() as the only FIPS allowed DH parameters are the known safe primes now (aren't they?).

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

paulidale commented:
With DH going legacy and key generation and signing no longer allowed, do we require this still?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

xnox commented:
> With DH going legacy and key generation and signing no longer allowed, do we require this still?

If it is in the submitted module, it is still needed.
And yes however bogus, the recalculation/rechecking of the public key is a valid selftest.

I wonder if this can at least go in as is, but under like OPENSSL_PEDANTIC_SELFTEST

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

paulidale reviewed (commented):
There needs to be some wrapping with the self testing condition: `ossl_fips_self_testing()`.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 247 days ago

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR has been closed.  It was waiting for the creator to make requested changes but it has not been updated for 297 days.
